src/net.cpp

```diff
@@ -1925,6 +1935,22 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
         int nTries = 0;
         while (!interruptNet)
         {
+            // If we didn't find an appropriate destination after trying 100 addresses fetched from addrman,
+            // stop this loop, and let the outer loop run again (which sleeps, adds seed nodes, recalculates
+            // already-connected network ranges, ...) before trying new addrman addresses.
+            nTries++;
+            if (nTries > 100)
+                break;
+
+            if (anchor) {
+                const CAddress addr = m_anchors.back();
+                m_anchors.pop_back();
+                if (!addr.IsValid() || IsLocal(addr) || !IsReachable(addr)) continue;
+                addrConnect = addr;
+                LogPrint(BCLog::NET, "Trying to make an anchor connection to %s\n", addrConnect.ToString());
+                break;
```

Comment:
I think the relation between connections opening order (anchors -> regular outbounds) and the enforcement of peer diversity should be underscored better. With this patchset, I believe outbound peer diversity is enforced by the fact that OUTBOUND_FULL_RELAY/BLOCK_RELAY are opened after anchors and thus those regular outbound addresses will be checked against the anchor ones. 

It could be better to exist this loop normally to let other checks happening while avoiding to call AddrMan::Select thanks to  the new`anchor` flag ? And add a comment above the connection type selection loop underscoring the order importance to avoid future work modifying it. E.g opening an OUTBOUND_FULL_RELAY may evict inadvertently an anchor in the same netgroup, thus diminishing effectiveness of anchors.

It would also enforce `HasDesirableServiceFlags()` on anchors. It's unlikely we learn change of their services before we connect to them but our desirability may have changed between a restart. Further, in case of a buggy/malicious `anchors.dat` we should check network groups distinctness of anchors among themselves.

Reply:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/17428#issuecomment-700159056).

Keeping the added code in its own block for the following reasons:
- some checks in the loop require `CAddrInfo` type but `m_anchors` contains only `CAddress` instances
- smaller diff